numpy/core/_internal.py

```diff
@@ -355,15 +355,38 @@ def _view_is_safe(oldtype, newtype):
         If the new type is incompatible with the old type.
 
     """
+    from numpy.lib.type_check import find_dtype_offsets
 
     # if the types are equivalent, there is no problem.
     # for example: dtype((np.record, 'i4,i4')) == dtype((np.void, 'i4,i4'))
     if oldtype == newtype:
         return
 
-    if newtype.hasobject or oldtype.hasobject:
-        raise TypeError("Cannot change data-type for object array.")
-    return
+    # no object members is fine
+    if not newtype.hasobject and not oldtype.hasobject:
+        return
```

Comment:
I don't think you meant to comment on this line, because it matches the old behaviour here.

I think I see the problem now. It doesn't exist yet, right, but would be a consequence of allowing partial views.

But as it stands, this pr doesn't allow the creation of those partial views in the first place. I believe this pr allows a strict subset of the things a complete partial-view implementation would allow